File: backend/app/provider/kakuyomu.py
import requests
import bs4

from app.provider.base import BookProvider
from app.model import (
    Author,
    TocEpisodeToken,
    TocChapterToken,
    BookMetadata,
    Episode,
)


class Kakuyomu(BookProvider):
    provider_id = "kakuyomu"
    lang = "jp"
    session = requests.Session()

    def _get_book_metadata(self, book_id: str) -> BookMetadata:
        url = f"https://kakuyomu.jp/works/{book_id}"
        res = self.session.get(url)
        res.raise_for_status()
        soup = bs4.BeautifulSoup(res.text, "html.parser")

        book_title = soup.find("h1", {"id": "workTitle"}).find("a").text

        authors = [
            Author(name=a.text, link=f"https://kakuyomu.jp{a['href']}")
            for a in soup.find("span", {"id": "workAuthor-activityName"}).findAll("a")
        ]

        introduction = soup.find("p", {"id": "introduction"}).text

        toc = []
        for li_tag in soup.find("ol", {"class": "widget-toc-items"}).find_all("li"):
            class_attr = li_tag["class"]
            if "widget-toc-chapter" in class_attr:
                title = li_tag.find("span").text
                if "widget-toc-level1" in class_attr:
                    toc.append(TocChapterToken(title=title, level=1))
                elif "widget-toc-level2" in class_attr:
                    toc.append(TocChapterToken(title=title, level=2))
                else:
                    raise RuntimeError("parse error")
            elif "widget-toc-episode" in class_attr:
                title = li_tag.find("span").text
                episode_id = li_tag.find("a")["href"].split("/")[-1]
                toc.append(TocEpisodeToken(title=title, episode_id=episode_id))
            else:
                raise RuntimeError("parse error")

        return BookMetadata(
            title=book_title,
            authors=authors,
            introduction=introduction,
            toc=toc,
        )

    def _get_episode(
        self,
        book_id: str,
        episode_id: str,
    ) -> Episode:
        url = f"https://kakuyomu.jp/works/{book_id}/episodes/{episode_id}"
        res = self.session.get(url)
        res.raise_for_status()
        soup = bs4.BeautifulSoup(res.text, "html.parser")

        paragraphs = [
            p.text
            for p in soup.find("div", {"class": "widget-episodeBody"}).find_all("p")
        ]

        return Episode(paragraphs=paragraphs)

File: src/main/kotlin/data/provider/providers/Kakuyomu.kt
package data.provider.providers

import data.provider.*
import io.ktor.client.request.*

class Kakuyomu : WebNovelProvider {
    companion object {
        const val id = "kakuyomu"
    }

    override suspend fun getRank(options: Map<String, String>): List<RemoteNovelListItem> {
        TODO("Not yet implemented")
    }

    override suspend fun getMetadata(novelId: String): RemoteNovelMetadata {
        val url = "https://kakuyomu.jp/works/$novelId"
        val doc = client.get(url).document()

        val title = doc
            .selectFirst("h1#workTitle > a")!!
            .text()

        val author = doc
            .selectFirst("span#workAuthor-activityName > a")!!
            .selectFirst("a")!!
            .let {
                RemoteNovelMetadata.Author(
                    name = it.text(),
                    link = "https://kakuyomu.jp" + it.attr("href"),
                )
            }

        val introduction = doc
            .selectFirst("p#introduction")
            ?.wholeText()
            ?.trimEnd()
            ?.removeSuffix("…続きを読む")
            ?: ""

        val toc = doc
            .select("ol.widget-toc-items > li")
            .map {
                RemoteNovelMetadata.TocItem(
                    title = it.selectFirst("span")!!.text(),
                    chapterId = it.selectFirst("a")?.attr("href")?.substringAfterLast("/")
                )
            }

        return RemoteNovelMetadata(
            title = title,
            authors = listOf(author),
            introduction = introduction,
            toc = toc,
        )
    }

    override suspend fun getChapter(novelId: String, chapterId: String): RemoteChapter {
        val url = "https://kakuyomu.jp/works/$novelId/episodes/$chapterId"
        val doc = client.get(url).document()
        return RemoteChapter(paragraphs = doc.select("div.widget-episodeBody > p").map { it.text() })
    }
}